### ISSUE TITLE
Is the implementation of `MultipleNegativesRankingLoss` right?

### ISSUE THREAD

**Issue by Ywandung-Lyou** [UKPLab/sentence-transformers#2521]
It is confusing why the labels are `range(len(scores))`.
```python
class MultipleNegativesRankingLoss(nn.Module):
    def __init__(self, model: SentenceTransformer, scale: float = 20.0, similarity_fct=util.cos_sim):
        super(MultipleNegativesRankingLoss, self).__init__()
        self.model = model
        self.scale = scale
        self.similarity_fct = similarity_fct
        self.cross_entropy_loss = nn.CrossEntropyLoss()

    def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor):
        reps = [self.model(sentence_feature)["sentence_embedding"] for sentence_feature in sentence_features]
        embeddings_a = reps[0]
        embeddings_b = torch.cat(reps[1:])

        scores = self.similarity_fct(embeddings_a, embeddings_b) * self.scale
        labels = torch.tensor(
            range(len(scores)), dtype=torch.long, device=scores.device
        )  # Example a[i] should match with b[i]
        return self.cross_entropy_loss(scores, labels)

    def get_config_dict(self):
        return {"scale": self.scale, "similarity_fct": self.similarity_fct.__name__}
```

**Comment by tomaarsen** [comment]
Hello!

The implementation indeed looks quite odd, especially because `labels` is overridden by this weird `range(len(scores))`, but the implementation is actually correct. Let's first discuss what happens above.

1. `sentence_features` becomes one list for each sentence given to the `InputExample`. For MNRL this is usually either 2 or 3, for (anchor, positive) pairs and for (anchor, positive, negative) triplets, as the [documentation](https://sbert.net/docs/package_reference/losses.html#multiplenegativesrankingloss) shows. In [this example](https://github.com/UKPLab/sentence-transformers/blob/66e0ee30843dd411c64f37f65447bb38c7bf857a/examples/training/nli/training_nli_v2.py#L85), it's 3 lists.
2. `reps` thus becomes a list of length 2 or 3, where each element is a tensor of size `[batch_size, output_dimension]`, e.g. `[128, 768]` by default in that example script.
3. `embeddings_a` thus becomes a tensor of `[128, 768]` representing the embeddings for the anchor sentences
4. `embeddings_b` becomes a tensor of `[256, 768]` representing the embeddings of the positive sentences concatenated with the negative sentences.
5. We compute the `scores` by taking the similarity between the two embedding tensors, which results in a scores matrix of shape `[128, 256]`. This is because for each of the 128 anchor embeddings, we compared against the 256 positive + negative embeddings. So, we have these `[128, 256]` scores.
6. Now, given these scores, we want the "anchor" embeddings to be very similar to the corresponding "positive" embedding, but different from 1) the corresponding negative embedding, 2) the non-corresponding positive embeddings and 3) the non-corresponding negative embeddings. So let's simplify this a little bit, and only consider the very first (anchor, positive, negative) triplet. For the anchor sentence, we have `256` scores (i.e., `scores[0]`), i.e. to **all** positive & negative sentences.
The goal for this training step is to have the score for the positive pair to be high, and all others to be low, so we want the score for the first anchor with the first positive to be high, i.e. `scores[0][0]` must be high while all other 255 scores are low.
We can repeat this process for any of the 128 triplets. E.g. if we take the triplet with index 80, then we again have 256 scores (represented by `scores[80]`), and now the score for the anchor-positive pair is at `scores[80][80]`.
7. So, we always want the scores on the diagonal of `scores` to be high, and all other scores to be low. We can do this by setting the labels as the `range(len(scores))` and using a cross-entropy loss. The CE loss will kind of see this as a classification: for each of the 128 samples, there are 256 potential options (or 'classes'), and one 'class' is correct (only the positive pair). All others are incorrect, and so the model learns to bring the embeddings of positive pairs together & push the embeddings of negative pairs apart.

I hope that clarifies it a bit - I understand that it might still be somewhat confusing. Another recommendation is to clone Sentence Transformers, add `breakpoint()` to the MNRL forward method & run https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/nli/training_nli_v2.py locally. Then you can, in your terminal, print out the shapes etc. of all parameters for a better understanding of what everything means.

- Tom Aarsen

**Comment by Ywandung-Lyou** [comment]
Thanks for your detailed explanation and I get it.
However, here I meet a problem. I am training an embedding model with the base of bert-base-zh with the self-made dataset. Each tuple is (anchor, positive). The performance of embeddings is judged by how many tuple whose cosine score is above 0.8. I find that the more I train, the worse the performance is. And the original pretrained model works the best.
From your view, what might be the reason?
